### PR TITLE
Add Heii On-Call Notification Provider

### PR DESCRIPTION
**Pull request by hevans66**
Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

Add notification provider for [Heii On-Call](https://heiioncall.com/).

## Type of change

Please delete any options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

### Setup Notification

image

### Events

| Event | Before | After |
|-------|------|-----|
| `UP` | - |  |
| `DOWN` | - |  |
| Certificate-expiry | - |  |
| Testing | - |  |

### CODE REVIEW

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Mostly, this is looking good.
I have left some inline remarks about UX and code-quality.

Please attach screenshots of the 3 types of events `DOWN`/`UP`, `Certificate Expiry`[^1] and `Testing` to the description.
(we should really document this one in the contribution guide...)

[^1]: See https://badssl.com/ for broken certs

File: src/components/notifications/HeiiOnCall.vue [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        <label for="heiioncall-apikey" class="form-label">{{ $t("API Key") }}<span
+            style="color: red;"
+        ><sup>*</sup></span></label>
+        <HiddenInput
+            id="heiioncall-apikey" v-model="$parent.notification.heiiOnCallApiKey" required="true"
+            autocomplete="false"
+        ></HiddenInput>
+    </div>
+    <div class="mb-3">
+        <label for="heiioncall-trigger-id" class="form-label">{{ $t("Trigger ID") }}<span
```
[the remark this answers is not in the document]
Comment (hevans66) [reply src/components/notifications/HeiiOnCall.vue:12]: Not a stupid question at all. I feel like this is one of those that can go either way. I suppose in this context we are using Trigger as basically a proper noun. I'll remove the tranlation.

File: server/notification-providers/heii-oncall.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        // Payload to Heii On-Call is the entire heartbat JSON
+        const payload = heartbeatJSON ? heartbeatJSON : {};
+
+        // If we can, add url back to monitor to payload
+        const baseURL = await setting("primaryBaseURL");
+        if (baseURL && monitorJSON) {
+            payload["url"] = baseURL + getMonitorRelativeURL(monitorJSON.id);
+        }
+
+        if (!heartbeatJSON) {
```
[the remark this answers is not in the document]
Comment (hevans66) [reply server/notification-providers/heii-oncall.js]: I believe my changes do what you asked for. Resolving this comment, but let me know if what I did missed the mark.

File: src/components/notifications/HeiiOnCall.vue [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            style="color: red;"
+        ><sup>*</sup></span></label>
+        <HiddenInput
+            id="heiioncall-trigger-id" v-model="$parent.notification.heiiOnCallTriggerId" required="true"
+            autocomplete="false"
+        ></HiddenInput>
+    </div>
+    <div class="form-text">
+        <i18n-t tag="p" keypath="wayToGetHeiiOnCallDetails" style="margin-top: 8px;">
+            <a href="https://heiioncall.com/docs" target="_blank">Heii On-Call Documentation</a>
```
[the remark this answers is not in the document]
Comment (hevans66) [reply src/components/notifications/HeiiOnCall.vue]: Trigger ID is very front and center as soon you create a trigger in Heii On-Call. It is mentioned in many places on Manual Trigger type documentation. I don't think there will be any confusion as to where to grab that from. If you would like I can add more text here to be explicit about how to find the Trigger ID, but I figured we wanted to keep the Setup Notification screen as clean as possible.

**Comment by hevans66** [comment]
@CommanderStorm Thanks for the review. I believe I have addressed all your concerns.

File: src/components/notifications/HeiiOnCall.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/components/notifications/HeiiOnCall.vue]: As long as how would be clear to a new user of said software, this is fine by me ^^

File: server/notification-providers/heii-oncall.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        // Post to Heii On-Call Trigger https://heiioncall.com/docs#manual-triggers
+        await axios.post(
+            `https://heiioncall.com/triggers/${notification.heiiOnCallTriggerId}/${action}`,
+            payload,
+            config
+        );
+
+        return "Sent Successfully";
+    }
```
Comment (CommanderStorm) [inline server/notification-providers/heii-oncall.js:64]: This should refactor the code to be closer to what the other notification providers look like => keeping it more consistent.

```suggestion
    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
        const okMsg = "Sent Successfully.";
        const payload = heartbeatJSON || {};

        const baseURL = await setting("primaryBaseURL");
        if (baseURL && monitorJSON) {
            payload["url"] = baseURL + getMonitorRelativeURL(monitorJSON.id);
        }

        const config = {
            headers: {
                Accept: "application/json",
                "Content-Type": "application/json",
                Authorization: "Bearer " + notification.heiiOnCallApiKey,
            },
        };
        
        const heiiUrl = `https://heiioncall.com/triggers/${notification.heiiOnCallTriggerId}/`;
        
        // docs https://heiioncall.com/docs#manual-triggers
        try {
            if (!heartbeatJSON) {
                // Testing or general notification like certificate expiry
                payload["msg"] = msg;
                await axios.post(heiiUrl + "alert", payload, config);
                return okMsg;
            }

            if (heartbeatJSON.status === DOWN) {
                await axios.post(heiiUrl + "alert", payload, config);
                return okMsg;
            }
            if (heartbeatJSON.status === UP) {
                await axios.post(heiiUrl + "resolve",payload, config);
                return okMsg;
            }
        } catch (error) {
            this.throwGeneralAxiosError(error);
        }
    }
```

File: src/components/notifications/HeiiOnCall.vue [hunk 5]
```diff
@@ -9,19 +9,17 @@
...
             autocomplete="false"
         ></HiddenInput>
     </div>
-    <div class="form-text">
-        <i18n-t tag="p" keypath="wayToGetHeiiOnCallDetails" style="margin-top: 8px;">
-            <a href="https://heiioncall.com/docs" target="_blank">Heii On-Call Documentation</a>
-        </i18n-t>
-    </div>
+    <i18n-t tag="p" keypath="wayToGetHeiiOnCallDetails" class="form-text mt-3">
+        <a href="https://heiioncall.com/docs" target="_blank">{{ $t("documentationOf", ["Heii On-Call"]) }}</a>
```
Comment (CommanderStorm) [inline src/components/notifications/HeiiOnCall.vue:21]: I think adding translations in this way might be easier for our translators

```suggestion
        <template #documentation>
            <a href="https://heiioncall.com/docs" target="_blank">{{ $t("documentationOf", ["Heii On-Call"]) }}</a>
        </template>
```

File: src/lang/en.json [hunk 6]
```diff
@@ -884,6 +884,6 @@
     "GrafanaOncallUrl": "Grafana Oncall URL",
     "Browser Screenshot": "Browser Screenshot",
     "What is a Remote Browser?": "What is a Remote Browser?",
-    "Trigger ID": "Trigger ID",
-    "wayToGetHeiiOnCallDetails": "To get a Heii On-Call Tigger ID and Api Token visit: {0}"
+    "wayToGetHeiiOnCallDetails": "How to get the Tigger ID and API Keys is explained in the {0}",
```
Comment (CommanderStorm) [inline src/lang/en.json:887]:
```suggestion
    "wayToGetHeiiOnCallDetails": "How to get the Tigger ID and API Keys is explained in the {documentation}",
```

File: src/components/notifications/HeiiOnCall.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/components/notifications/HeiiOnCall.vue]: resolving this one, as below is better to track this (below has the fixed indentation ^^)

File: server/notification-providers/heii-oncall.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        const config = {
+            headers: {
+                Accept: "application/json",
+                "Content-Type": "application/json",
+                Authorization: "Bearer " + notification.heiiOnCallApiKey,
+            },
+        };
+
```
Comment (CommanderStorm) [inline server/notification-providers/heii-oncall.js:28]:
```suggestion

```

File: server/notification-providers/heii-oncall.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        const config = {
+            headers: {
+                Accept: "application/json",
+                "Content-Type": "application/json",
+                Authorization: "Bearer " + notification.heiiOnCallApiKey,
+            },
+        };
+        
+        const heiiUrl = `https://heiioncall.com/triggers/${notification.heiiOnCallTriggerId}/`;
+
```
Comment (CommanderStorm) [inline server/notification-providers/heii-oncall.js:30]:
```suggestion

```

File: server/notification-providers/heii-oncall.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                await axios.post(heiiUrl + "alert", payload, config);
+                return okMsg;
+            }
+
+            if (heartbeatJSON.status === DOWN) {
+                await axios.post(heiiUrl + "alert", payload, config);
+                return okMsg;
+            }
+            if (heartbeatJSON.status === UP) {
+                await axios.post(heiiUrl + "resolve",payload, config);
```
Comment (CommanderStorm) [inline server/notification-providers/heii-oncall.js:45]:
```suggestion
                await axios.post(heiiUrl + "resolve", payload, config);
```

**Review by CommanderStorm — COMMENTED** [review]
formatting issues introduced in above fix

**Review by CommanderStorm — APPROVED** [review]
Okay, I changed the formatting of the code a bit to be closer to the other notification providers.
Could you review the changes I added in https://github.com/louislam/uptime-kuma/pull/4485/commits/cc000117ca9b8f888c69a31b9c8f2841b79bab7f?

Other than that, I am happy with the current code.

I am unsure if we should change the translations in this context as suggested in https://github.com/louislam/uptime-kuma/pull/4485#pullrequestreview-1881439791 => I am going to leave this up to you to either accept/resolve

**Comment by hevans66** [comment]
@CommanderStorm awesome! Accepted your changes. Thanks so much for the review.

**Comment by hevans66** [comment]
@CommanderStorm let me know if there is anything else I can do to help this get merged. I'd like to work on a guide for setting up Uptime Kuma with Heii On-Call, and would like to have this in before its published.

**Comment by CommanderStorm** [comment]
Actually there is: 
I missed in the previous review that this is missing the screenshot from the `Testing` state (the "Test"-button in above screenshot)

Other than that I would be happy to merge this into the upcoming v2.0 release.

**Comment by hevans66** [comment]
Testing screenshot has been added.

**Comment by CommanderStorm** [comment]
Thanks for the notification provider!
I think adding it is uncontroversial as it matches the code style and the appropriate tests have been performed.
⇒ merging with junior maintainer approval.

**Pull request merged**